### ISSUE TITLE
Embedded broker as maven dependency

### ISSUE THREAD

**Issue by dowhiletrue** [hivemq/hivemq-community-edition#19]
## Problem or use case
I would like to test my mqtt client/broker setup using JUnit tests.
Since it is cumbersome to start a broker before running the tests, an embedded MQTT 5 compliant broker would ease up the testing.

## Preferred solution or suggestions
* HiveMQ broker as maven dependency
* Documentation on how to start the broker within a running Java process (e.g. Junit test)

**Comment by sauroter** [comment]
Hi @dowhiletrue,
Thank you for your feature request.

We see the need for something like that, but are right now not sure how to best tackle the issue. Internal considerations are ongoing at the moment.

We will keep the issue open and track progress once decisions are made.

Kind regards
Georg

**Comment by siepkes** [comment]
@sauroter I see some places (like [this one](https://www.hivemq.com/blog/hivemq-1-4-3-released/)) where an embedded mode is mentioned. However I can't find any documentation about it. Does the embedded mode still exist?

**Comment by ObviousDWest** [comment]
I am also interested in this feature. I would prefer it be compatible with Spring (or at least not preclude it), so I can integrate with other standard systems (logging, actuator, RestTemplate, caching, permissions, spring security, maybe jms...). Eg. for implementing hivemq extensions. While some of this can be done by splitting things into microservices, integration tests are much easier if I can recombine all that into a compound application.

I guess I could modify HiveMQServer.java.

**Comment by abonstu** [comment]
Any further feedback here guys?

There are a bunch of projects looking to transition away from moquette but maven dependency support would seem to be a pretty common prerequisite.

**Comment by St0n3Gh0st** [comment]
I also need to be able to use maven to get HiveMQ CE as an embedded broker. 

Any timeline for either maven or gradle support?

**Comment by nhitchins** [comment]
Having the same problem - been looking for a solution for a while. Any ETA?

**Comment by ramod8** [comment]
Has there been any progress on this? I have been waiting on this for a while.

**Comment by sauroter** [comment]
Hi,
oops i forgot to update this ticket. I am sorry.
Actually there has been some progress. Wich at least in part solves this issue.

There now exists an official docker image for [HiveMQ CE](https://hub.docker.com/r/hivemq/hivemq-ce).

Now you can use the [Testcontainers](https://www.testcontainers.org/) project to pretty easily build a test dependency. Your custom container can look something like this:

```
public class HiveMQCEContainer extends GenericContainer<HiveMQCEContainer> {

    public HiveMQCEContainer() {
        super("hivemq/hivemq-ce:2020.2");

        this.withExposedPorts(1883)
                .waitingFor(Wait.forListeningPort());
    }


    @Test
    public void test() {
        final HiveMQCEContainer testContainer = new HiveMQCEContainer();
        testContainer.start();
        final Mqtt5BlockingClient client =
                Mqtt5Client.builder()
                        .serverHost("localhost")
                        .serverPort(testContainer.getMappedPort(1883))
                        .build().toBlocking();


        final Mqtt5ConnAck connAck = client.connect();

        assertEquals(0, connAck.getReasonCode().getCode());
    }
}
```
I know that this does not solve every part of this issue, but it should be a starting point.

Feel free to contact me if you have any questions.
Cheers

Georg

**Comment by fbeister** [comment]
Thanks for the heads-up - at least for my usecase (https://community.hivemq.com/t/hivemq-ce-as-maven-dependency/39) docker is not a solution. Should I open a separate issue for that? Do you have a roadmap for real maven integration?

**Comment by splatch** [comment]
Well, I know its kind of side topic, however as you ask for maven centric solution I can bring one idea. ActiveMQ 5.x can be embedded with [dedicated maven plugin](https://activemq.apache.org/maven2-activemq-broker-plugin.html) and it has support for [MQTT transport](https://activemq.apache.org/mqtt.html). It will be able to serve you as long as you need basic transport and not HiveMQ specific extensions on other side.

**Comment by sauroter** [comment]
We now have an embedded broker api in master, so this issue is closed.
